### ISSUE TITLE
What's the best way to disable a residual block?

### ISSUE THREAD

**Issue by versatran01** [ceres-solver/ceres-solver#700]
My situation is that I want to add a bunch of residual blocks to a problem instance and reuse it for multiple solves. 
This is like an ICP problem where you have to recompute data association after parameters change, which is a non-stationary NLS problem. 
Between each solve, after the new data association, I want to disable/enable some of the already added residual blocks depends on the results.
What is the best way to do this? Is it enough to just return true/false in the cost function's Evaluate() call to indicate whether this residual block takes part in the final solve?
Also to what extent can one reuse a problem instance? If I know in advance the max number of residual blocks, can I just have a persistent problem instance with preallocated residuals and modify them as I go?

Thanks!

**Comment by sandwichmaker** [comment]
Returning false will not work as it will cause the whole evaluation to fail. Why not just add and remove residuals as needed?
No you cannot pre-allocate the a max number of residuals. What are you trying to optimize ? Problem construction is quite fast and usually a negligible component of the actual solve time.

**Comment by versatran01** [comment]
If returning false is a no-go, then would it make sense to set residual to 0 in Evaluate()?

My problem is fairly small (around 400 residual blocks with 2 parameter blocks), my timing shows that building then problem takes around 300us and solve takes around 1ms (for 2~3 iterations).
Because it's an ICP problem it needs to redo the data association after each solve, which means it needs to rebuild the problem multiple times.

**Comment by sandwichmaker** [comment]
sure you can set the residual AND Jacobian to zero. But it will increase
the size of the problem that you are solving.
If your problem is so small, then why not just have a single costfunction?
since your jacobian is completely dense, you will actually be better off
doing that then having multiple cost functions and having that overhead.

Sameer

**Comment by versatran01** [comment]
Just want to clarify, when you say have a single cost function, do you mean a single residual block?
So instead of having multiple residual blocks (each corresponding to one point-to-point match), just have a single residual block that contains all the matches and determine which one to disable inside a for loop in Evaluate()?

**Comment by sandwichmaker** [comment]
Yeah. This will work if you do not have a loss function.

**Comment by versatran01** [comment]
Thanks. Since the GitHub email didn't include my later edits, I just want to ask again that if I combine all residual blocks (N) into one, I should also increase the num_residuals of the new cost function by N times? or is it ok to accumulate each one using an in-place operator like r[0] += r_i?

This also precludes multi-threaded evaluation of Jacobian, right?

**Comment by sandwichmaker** [comment]
Yes you need to increase the number of residuals. Do not accumulate. You
can do your own threading if you need it.
